khook/x86/hook.c

```diff
@@ -1,22 +1,112 @@
 	"mov %%cr0, %%rax\n\t"			\
 	"and $0xfffffffffffeffff, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	::: "%rax" )
 
-#define kernel_write_leave() asm volatile (	\
+# define kernel_write_leave() asm volatile (	\
 	"mov %%cr0, %%rax\n\t"			\
 	"or $0x0000000000010000, %%rax\n\t"	\
 	"mov %%rax, %%cr0\n\t"			\
 	"sti\n\t"				\
 	::: "%rax" )
-
+#endif
+#ifdef USE_PTE_FAM
+/* Because of syncronization mechanism
+ * this set_rw/set_ro can be used only pairy
+ * and should be used in small time window
+ **/
+static int
+	is_first_rw = 0x00,
+	is_last_rw = 0x00;
+static volatile int lock_pte = 0x00;
+/*|SPIN FAMILY|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//**/
+/* argument MUST take address to global volatile int */
+/*|*/static void lock(volatile int *spin__) {
+/*|*/    while (__sync_lock_test_and_set(spin__, 1)) { ; }
+/*|*/}
+/*|*/static void unlock(volatile int *spin__) {
+/*|*/    __sync_synchronize(); /* Memory barrier. */
+/*|*/    *spin__ = 0x00;
+/*|*/}
+/*|PTE FAMILY|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*/
+/*|*/static void set_addr_rw(volatile uint64_t addr, int *flag){
+/*|*/   unsigned int level;
+/*|*/   uint64_t *pte = NULL;
+/*|*/   pte = (uint64_t*)lookup_address((unsigned long)addr, &level);
+/*|*/   if (pte[0x00] & ~((unsigned long long int)1<<1))
+/*|*/		pte[0x00] |= ((unsigned long long int)1<<1), *flag = 0x00;
+/*|*/	else
+/*|*/		*flag = 0x01;
+/*|*/}
+/*|*/static void set_addr_ro(volatile uint64_t addr, int *flag){
+/*|*/   unsigned int level;
+/*|*/   uint64_t *pte = NULL;
+/*|*/   pte = (uint64_t*)lookup_address((unsigned long)addr, &level);
+/*|*/	if ( !*flag )
+/*|*/   	pte[0x00] = pte[0x00] & ~((unsigned long long int)1<<1);
+/*|*/}
+/*|*/static void set_addr_ex(volatile uint64_t addr){
+/*|*/   unsigned int level;
+/*|*/   uint64_t *pte = NULL;
+/*|*/   pte = (uint64_t*)lookup_address((unsigned long)addr, &level);
+/*|*/   if (pte[0x00] & ((unsigned long long int)1<<63))
+/*|*/           pte[0x00] ^= ((unsigned long long int)1<<63);
+/*|*/   __asm__("cpuid  \n\t");
+/*|*/}
+/*|*/static void set_addr_nx(volatile void *addr){
+/*|*/   unsigned int level;
+/*|*/   uint64_t *pte = NULL;
+/*|*/   pte = (uint64_t*)lookup_address((unsigned long)addr, &level);
+/*|*/   if ( !(pte[0x00] & ((unsigned long long int)1<<63)))
+/*|*/           pte[0x00] ^= ((unsigned long long int)1<<63);
+/*|*/   __asm__("cpuid  \n\t");
+/*|*/}
+/*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*//*|*/
 ////////////////////////////////////////////////////////////////////////////////
 // IN-kernel length disassembler engine (x86 only, 2.6.33+)
 ////////////////////////////////////////////////////////////////////////////////
+#endif /* USE_PTE_FAM */
+
+#define MAKE_RW(addr) make_rw((uint64_t)addr)
+static void make_rw(uint64_t kaddr){
+	#ifdef USE_CR0_FAM
+	        kernel_write_enter();
+	#endif
+	#ifdef USE_PTE_FAM
+		lock( &lock_pte );
+	        set_addr_rw( kaddr, &is_first_rw );
```

Comment:
not sure the approach is correct ... having in mind you'll be having `ptr` and `len` as arguments, could you rewrite the implementation to iterate all PTE's inside the range and change their flags accordingly?

Reply:
If I got you right - you want me adding flags for pages to hook struct? 
I think tn this case we can save all used page's PTE entris, full. Previously reserve memory for, in worst case, two entries. And restore it on finit.